File: Game.js
class Game {
    constructor(params) {
        this.inning=1;
        this.outs=0;
        this.awayOrHome=0;
        this.bases=[0,0,0];
        this.score=[0,0];
        this.gameOn=true;
        this.setGameState(params);
    }
    
    atBat() {
        const outcome = Math.random() * 186518;
        if (outcome < 126600) return 'Out';
        else if (outcome < 144479) return 'BB';
        else if (outcome < 170426) return '1B';
        else if (outcome < 178957) return '2B';
        else if (outcome < 179742) return '3B';
        else return 'HR';
    }
    
    arraySum(array) {
        return array.reduce((sum,element)=>sum+element,0);
    }
    
    walker() {
        this.setGameState({bases: this.bases.concat(0)});
        this.setGameState({bases: this.bases.map((elm,i)=>i===0? elm+1:elm)});
        for (let i = 0; i < 3; i++) {
            if (this.bases[i]==2) {
                this.setGameState({
                    bases: this.bases.map((elm,j)=>{
                        if (j===i) return this.bases[j] - 1;
                        else if (j===i+1) return this.bases[j] + 1;
                        else return this.bases[j];
                    })
                })
            }
        }
        const runs = this.bases[this.bases.length-1];
        this.setGameState({
            bases: this.bases.slice(0,3),
            score: this.score.map((elm,i)=>i===this.awayOrHome ? elm + runs : elm)
        });
    }
    
    hitter(hitType) {
        if (hitType === '1B') this.bases = [1,0].concat(this.bases);
        else if (hitType === '2B') this.setGameState({bases: [0,1].concat(this.bases)});
        else if (hitType === '3B') this.setGameState({bases: [0,0,1].concat(this.bases)});
        else if (hitType === 'HR') this.setGameState({bases: [0,0,0,1].concat(this.bases)});
        const runs = this.arraySum(this.bases.slice(3,this.bases.length));
        this.setGameState({
            bases: this.bases.slice(0,3),
            score: this.score.map((elm,i)=>i===this.awayOrHome ? elm + runs : elm)
        });
    }
    
    atBatResult() {
        const myAtBat = this.atBat();
        if (myAtBat === 'Out') this.setGameState({outs: this.outs + 1});
        else if (myAtBat === 'BB') this.walker();
        else this.hitter(myAtBat);
        if ((this.inning >= 9 && ((this.outs >= 3 && this.awayOrHome === 0) || this.awayOrHome === 1) && this.score[0] < this.score[1]) || 
        (this.inning >= 9 && this.outs >= 3 && this.awayOrHome === 1 && this.score[0] > this.score[1])) {
            this.setGameState({gameOn: false});
        }
        if (this.outs >= 3) {
            if (this.awayOrHome === 1) {
                this.setGameState({inning: this.inning + 1});
            }
            this.setGameState({
                outs: 0,
                awayOrHome: (this.awayOrHome + 1) % 2,
                bases: [0,0,0]
            })
        }
    }
    
    playGame() {
        while (this.gameOn) {
            this.atBatResult();
        }
        const finalScore = JSON.parse(JSON.stringify(this.score));
        const winner = this.score[0] > this.score[1] ? 0 : 1;
        this.setGameState({
            inning: 1,
            outs: 0,
            awayOrHome: 0,
            bases: [0,0,0],
            score: [0,0],
            gameOn: true
        });
        return {finalScore, winner};
    }
    
    setGameState(params) {
        Object.assign(this, params);
    }

}

File: python/classes/Game.py
import copy

class Game:
    def __init__(self,
                 teams,
                 inning=1,
                 outs=0,
                 away_or_home=0,
                 bases=[0,0,0],
                 score=[0,0],
                 game_on=True):
        self.teams=teams
        self.inning=inning
        self.outs=outs
        self.away_or_home=away_or_home
        self.bases=bases
        self.score=score
        self.game_on=game_on
        self.current_player=[0,0]

    def walker:
        self.bases.append(0)
        self.bases[0] += 1
        for i in range(3):
            if self.bases[i]==2:
                self.bases[i] -= 1
                self.bases[i+1] += 1
        runs = self.bases[-1]
        self.bases = self.bases[:3]
        self.score[self.away_or_home] += runs

    def hitter(self, hit_type):
        if hit_type == '1B':
            self.bases = [1,0].extend(self.bases)
        elif hit_type == '2B':
            self.bases = [0,1].extend(self.bases)
        elif hit_type == '3B':
            self.bases = [0,0,1].extend(self.bases)
        elif hit_type == 'HR':
            self.bases = [0,0,0,1].extend(self.bases)
        runs = sum(self.bases[3:])
        self.bases = self.bases[:3]
        self.score[self.away_or_home] += runs

    def at_bat(self):
        player=self.teams[self.away_or_home][self.current_player]
        result = player.at_bat()
        if result == 'OUT':
            self.outs += 1
        elif result == 'BB':
            self.walker()
        else:
            self.hitter(result)
        if (self.inning >= 9
            and ((self.outs >= 3 and self.away_or_home == 0)
            or self.away_or_home == 1)
            and self.score[0] < self.score[1])
            or (self.inning >= 9
            and self.outs >= 3
            and self.score[0] > self.score[1]):
            self.game_on = False
        if self.outs >= 3:
            if self.away_or_home == 1:
                self.inning += 1
            self.outs = 0
            self.current_player[self.away_or_home] = (self.current_player[self.away_or_home] + 1) % 9
            self.away_or_home = (self.away_or_home + 1) % 2
            bases = [0, 0, 0]

    def play_game(self):
        while self.game_on:
            self.at_bat()
        final_score = copy.copy(self.score)
        winner = 1 if (self.score[0] < self.score[1]) else 0
        self.inning = 1
        self.outs = 0
        self.away_or_home = 0
        self.bases = [0,0,0]
        self.score = [0,0]
        self.game_on = True
        return {
            "final_score": final_score,
            "winner": winner
        }